server/migrator/Exporter.java

```diff
@@ -0,0 +1,197 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package grakn.core.server.migrator;
+
+import grakn.core.Grakn;
+import grakn.core.common.exception.GraknException;
+import grakn.core.common.parameters.Arguments;
+import grakn.core.concept.thing.Attribute;
+import grakn.core.concept.thing.Entity;
+import grakn.core.concept.thing.Relation;
+import grakn.core.concept.thing.Thing;
+import grakn.core.concept.type.RoleType;
+import grakn.core.server.Version;
+import grakn.core.server.migrator.proto.DataProto;
+import grakn.core.server.migrator.proto.MigratorProto;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Stream;
+
+import static grakn.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+import static grakn.core.common.exception.ErrorMessage.Migrator.FILE_NOT_WRITABLE;
+
+public class Exporter implements Migrator {
+
+    private static final Logger LOG = LoggerFactory.getLogger(Exporter.class);
+    private final Grakn.Session session;
+    private final Path filename;
+    private long totalThingCount = 0;
+    private final AtomicLong entityCount = new AtomicLong(0);
+    private final AtomicLong relationCount = new AtomicLong(0);
+    private final AtomicLong attributeCount = new AtomicLong(0);
+    private final AtomicLong ownershipCount = new AtomicLong(0);
+    private final AtomicLong playerCount = new AtomicLong(0);
+
+    public Exporter(Grakn grakn, String database, Path filename) {
+        this.session = grakn.session(database, Arguments.Session.Type.DATA);
+        this.filename = filename;
+    }
+
+    @Override
+    public MigratorProto.Job.Progress getProgress() {
+        long current = attributeCount.get() + relationCount.get() + entityCount.get();
+        return MigratorProto.Job.Progress.newBuilder()
+                .setCurrent(current)
+                .setTotal(totalThingCount)
+                .build();
+    }
+
+    @Override
+    public void run() {
+        LOG.info("Exporting {} from Grakn {}", session.database().name(), Version.VERSION);
+        try (OutputStream outputStream = new BufferedOutputStream(Files.newOutputStream(filename))) {
+            try (Grakn.Transaction tx = session.transaction(Arguments.Transaction.Type.READ)) {
+                totalThingCount = tx.concepts().getRootThingType().getInstancesCount();
```

Comment:
Something worth getting into the habit of is marking as `final` local variables whose value never changes, such as this `totalThingCount`.

Not only does this make algorithms easier to parse (as we don't have to consider the possibility of variables being mutated further down the method) but it will have an additional benefit if/when we migrate to a stricter language such as Rust: in Rust, variables are immutable by default. So we will give special attention to variables that are not marked as `final` in Java.

This can actually be refactored very easily in IntelliJ using:

1. Analyze
2. Run Inspection by Name
3. Local variable or parameter can be final
4. Uncheck "Report method parameters"
5. Fix All

Reply:
`totalThingCount` is not a final variable, it's a field in class that will be read by the progress reporting thread.

Reply:
Do we have a strict rule on this? Having everything to be final is quite noisy.

Reply:
It is Java's fault that they chose to make variables non-final by default. Adding `final` to variables is definitely a workaround.

But it shouldn't be seen as noise - code implemented using final variables is usually a lot more readable!

Reply:
OK I finalized everything.